File: app/src/main/java/torille/fi/lurkforreddit/comments/CommentPresenterModule.java
package torille.fi.lurkforreddit.comments;

import dagger.Module;
import dagger.Provides;
import torille.fi.lurkforreddit.data.RedditRepository;
import torille.fi.lurkforreddit.data.models.view.Post;

/**
 * This is a Dagger module. We use this to pass in the View dependency to the
 * {@link CommentPresenter}.
 */
@Module
public class CommentPresenterModule {

    private final Post mPost;
    private final boolean mIsSingleCommentThread;

    public CommentPresenterModule(Post post, boolean isSingleCommentThread) {
        mPost = post;
        mIsSingleCommentThread = isSingleCommentThread;
    }

    @Provides
    CommentContract.Presenter<CommentContract.View> provideCommentContractPresenter(RedditRepository repository,
                                                                                    Post post,
                                                                                    boolean isSingleCommentThread) {
        return new CommentPresenter(repository, post, isSingleCommentThread);
    }

    @Provides
    Post provideCommentOriginalPost() {
        return mPost;
    }

    @Provides
    boolean provideIsSingleCommentThread() { return mIsSingleCommentThread;}
}

File: app/src/main/java/torille/fi/lurkforreddit/comments/CommentPresenterModule.kt
package torille.fi.lurkforreddit.comments

import dagger.Binds
import dagger.Module
import dagger.Provides
import dagger.android.ContributesAndroidInjector
import torille.fi.lurkforreddit.data.models.view.Post
import torille.fi.lurkforreddit.di.scope.ActivityScoped
import torille.fi.lurkforreddit.di.scope.FragmentScoped

@Module
abstract class CommentPresenterModule {

    @FragmentScoped
    @ContributesAndroidInjector
    internal abstract fun commentFragment(): CommentFragment

    @ActivityScoped
    @Binds
    internal abstract fun commentPresenter(presenter: CommentPresenter): CommentContract.Presenter

    @Module
    companion object {
        @JvmStatic
        @Provides
        @ActivityScoped
        internal fun provideCommentOriginalPost(commentActivity: CommentActivity): Post {
            return commentActivity.intent.getParcelableExtra(CommentActivity.EXTRA_CLICKED_POST)
        }

        @JvmStatic
        @Provides
        @ActivityScoped
        internal fun provideIsSingleCommentThread(commentActivity: CommentActivity): Boolean {
            return commentActivity.intent.getBooleanExtra(CommentActivity.IS_SINGLE_COMMENT_THREAD, false)
        }
    }
}
